### DISCUSSION TITLE
callback on record insert, delete and update

### DISCUSSION THREAD

**Discussion by sugizo** [sqlalchemy/sqlalchemy#5882] [Usage Questions] [answered]
# Objective
before insert table can  execute python function to convert data 
**e.g.** 
user insert Field intervals in list [2, 2, 1, 2, 2, 2, 1]
then base on list above automatically convert and insert into Field intervals_name with [W, W, H, W, W, W, H] on the same table

**or** 

after insert table can send email to user, base on field email stored on another table 
**e.g.**
```
def after_insert_nav(f, id):
    query_account_0 = (db.account.product == f['product'] )
    rows_account = db(query_account_0).iterselect()
    for row_account in rows_account:
        if row_account.investor.notification:
            mail_to = row_account.investor.email
            subject = 'NAV %s' % row_account.product.name
            messages = response.render('templates/mail/nav.html', 
                                               dict(row = row_account, row_nav = row_nav) )
            attachments = ''
            insert_mail_queue(mail_to, subject, messages, attachments)
```

# question
is it possible to use callback on record insert, delete and update in sqlalchemy ?

**ref**
http://web2py.com/books/default/chapter/29/06/the-database-abstraction-layer#callbacks-on-record-insert-delete-and-update

**n.b.**
objective above can be achieve in [pydal](https://github.com/web2py/pydal)

thanks

**Comment by zzzeek** [comment, marked as answer]
The SQLAlchemy ORM has [lots of events](https://docs.sqlalchemy.org/en/13/orm/events.html) that are tailored towards so-called "lifecycle" events like object inserted, etc.   Core also has [lots of events](https://docs.sqlalchemy.org/en/13/core/events.html) which are a little bit more statement oriented.  The use cases you have above are directly available with the ORM, when using Core you would need to write events that are inspecting statements.

Assuming you're using ORM, the "before" events you would use include [before_flush](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=before_flush#sqlalchemy.orm.events.SessionEvents.before_flush) and [before_insert](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=before_insert#sqlalchemy.orm.events.MapperEvents.before_insert).       If using an event like before_insert, you would apply your additional SQL operations on the connection object that's passed.   

For the "after" events, there is [after_insert](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=after_insert#sqlalchemy.orm.events.MapperEvents.after_insert), but I would not use that alone to send an email because you're in an uncommitted transaction that might fail.   [after_commit](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=after_insert#sqlalchemy.orm.events.SessionEvents.after_commit) would be best to actually do something permanent like sending an email, but you might need to use after_insert as well in order to catch the object like this:

```python
from sqlalchemy import event
from sqlalchemy import inspect

@event.listens_for(Session, "after_begin")
def setup_transaction_events(session, transaction, connection):
    session.info["send_emails"] = []

@event.listens_for(MyClass, "after_insert")
def object_inserted(mapper, connection, target):
    session = inspect(target).session
    session.info["send_emails"].append(target)    

@event.listens_for(Session, "after_commit")
def send_emails(session):
    send_emails = session.info["send_emails"]
    while send_emails:
       obj = send_emails.pop(0)
       email_sender.send_an_email("Object %s was inserted" % obj)

```

APIs used above include:

* [inspect](https://docs.sqlalchemy.org/en/13/core/inspection.html?highlight=inspect#sqlalchemy.inspect) - inspect the object

* [Session.info](https://docs.sqlalchemy.org/en/13/orm/session_api.html?highlight=session%20info#sqlalchemy.orm.session.Session.info) - dictionary that you can place anything you want inside of, associated with a Session

* [InstanceState](https://docs.sqlalchemy.org/en/13/orm/internals.html#sqlalchemy.orm.state.InstanceState) - use the .session attribute to get the session for that object

* [after_begin](https://docs.sqlalchemy.org/en/13/orm/events.html#sqlalchemy.orm.events.SessionEvents.after_begin) - setup a space to track things you want to do when the transaction ends

* [after_insert](https://docs.sqlalchemy.org/en/13/orm/events.html#sqlalchemy.orm.events.MapperEvents.after_insert) - intercept specific objects being inserted

* [after_commit](https://docs.sqlalchemy.org/en/13/orm/events.html#sqlalchemy.orm.events.SessionEvents.after_commit) - do things after a commit occurs like send emails

There are frameworks such as [zope.transaction](https://transaction.readthedocs.io/en/latest/) that include more built-in support for things like "send an email when transaction is complete", I searched for some that are not associated with Zope like for example for Flask but didn't find anything too direct.    Just don't send emails before the transaction is committed!
